File: 16/day16.py
#!/usr/bin/env python3

import re

from functools import reduce
from itertools import chain

try:
    from math import prod
except ImportError:
    def prod(iter):
        return reduce(lambda x, y: x * y, iter)


def main():
    with open("input", "r") as f:
        lines = [i.split("\n") for i in f.read().strip().split("\n\n")]

    rules, ticket, nearby = parse(lines)

    print(solve1(rules, nearby))
    print(solve2(rules, ticket, nearby))


def solve1(rules, nearby):
    r = list(reduce(set.union, map(set, rules.values())))
    return sum(i for i in chain(*nearby) if i not in r)


def solve2(rules, ticket, nearby):
    r = list(reduce(set.union, map(set, rules.values())))
    valid = [i for i in nearby if all(j in r for j in i)]

    column_choices = list()
    for col in zip(*valid):
        choices = [k for k, v in rules.items() if all(i in v for i in col)]
        column_choices.append(choices)

    departure = list()
    while any(i for i in column_choices):
        for i, c in enumerate(column_choices):
            if len(c) == 1:
                choice = c[0]
                for j in column_choices:
                    if choice in j:
                        j.remove(choice)
                if choice.startswith("departure"):
                    departure.append(i)
                break

    return prod(ticket[i] for i in departure)


def parse(lines):
    _rules, _tickets, _nearby = lines

    _rules = [i.split(": ") for i in _rules]
    _tickets = _tickets[1:]
    _nearby = _nearby[1:]

    rules = {k: list() for k, _ in _rules}
    for field, value in _rules:
        for m in re.finditer(r"(\d+)-(\d+)", value):
            rules[field] += range(int(m.group(1)), int(m.group(2)) + 1)

    tickets = list(map(int, ",".join(_tickets).split(",")))
    nearby = list(map(lambda i: list(map(int, i.split(","))), _nearby))

    return rules, tickets, nearby


if __name__ == "__main__":
    main()

File: 16/day16.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"regexp"
	"strconv"
	"strings"
)

func main() {
	strFile, err := read("input")

	if err != nil {
		os.Exit(1)
	}

	rules, ticket, nearby := parse(strFile)
	fmt.Println(solve1(rules, nearby))
	fmt.Println(solve2(rules, ticket, nearby))
}

func solve1(rules map[string][][]int, nearby [][]int) int {
	sum := 0

	for _, i := range nearby {
		for _, j := range i {
			if !inRules(rules, j) {
				sum += j
			}
		}
	}

	return sum
}

func solve2(rules map[string][][]int, ticket []int, nearby [][]int) int {
	valid := [][]int{}
	for _, i := range nearby {
		if ticketValid(rules, i) {
			valid = append(valid, i)
		}
	}

	choices := [][]string{}
	for i := 0; i < len(valid[0]); i++ {
		choice := []string{}
		for k, v := range rules {
			check := true
			for _, t := range valid {
				if !inRule(v, t[i]) {
					check = false
					break
				}
			}

			if check {
				choice = append(choice, k)
			}
		}
		choices = append(choices, choice)
	}

	departure := []int{}
	for !allEmpty(choices) {
		for i, c := range choices {
			if len(c) == 1 {
				choice := c[0]

				for j, c2 := range choices {
					if _, ok := contains(c2, choice); ok {
						choices[j] = remove(c2, choice)
					}
				}

				if strings.HasPrefix(choice, "departure") {
					departure = append(departure, i)
				}

				break
			}
		}
	}

	prod := 1
	for _, d := range departure {
		prod *= ticket[d]
	}
	return prod
}

func parse(s string) (map[string][][]int, []int, [][]int) {
	lines := [][]string{}
	delim := getNewline(s)

	for _, l := range strings.SplitN(strings.TrimSpace(s), delim+delim, -1) {
		lines = append(lines, strings.SplitN(l, delim, -1))
	}

	reg := regexp.MustCompile(`(\d+)-(\d+)`)
	rules := map[string][][]int{}
	for _, r := range lines[0] {
		split := strings.SplitN(r, ": ", 2)
		rules[split[0]] = [][]int{}
		matches := reg.FindAllStringSubmatch(split[1], -1)
		for _, m := range matches {
			ranges := []int{}
			low, err1 := strconv.Atoi(m[1])
			high, err2 := strconv.Atoi(m[2])
			if err1 != nil || err2 != nil {
				return nil, nil, nil
			}
			ranges = append(ranges, low)
			ranges = append(ranges, high)
			rules[split[0]] = append(rules[split[0]], ranges)
		}
	}

	ticket := []int{}
	for _, i := range strings.SplitN(lines[1][1], ",", -1) {
		n, err := strconv.Atoi(i)
		if err != nil {
			return nil, nil, nil
		}
		ticket = append(ticket, n)
	}

	nearby := [][]int{}
	for _, l := range lines[2][1:] {
		nearbyTicket := []int{}
		for _, i := range strings.SplitN(l, ",", -1) {
			n, err := strconv.Atoi(i)
			if err != nil {
				return nil, nil, nil
			}
			nearbyTicket = append(nearbyTicket, n)
		}
		nearby = append(nearby, nearbyTicket)
	}

	return rules, ticket, nearby
}

func read(path string) (string, error) {
	bfile, err := ioutil.ReadFile(path)

	if err != nil {
		return "", err
	}

	file := string(bfile)
	return file, err
}

func getNewline(s string) string {
	for _, v := range s {
		if v == '\r' {
			return "\r\n"
		}
	}
	return "\n"
}

func ticketValid(r map[string][][]int, t []int) bool {
	for _, i := range t {
		if !inRules(r, i) {
			return false
		}
	}
	return true
}

func inRules(r map[string][][]int, n int) bool {
	for _, v := range r {
		if inRule(v, n) {
			return true
		}
	}
	return false
}

func inRule(r [][]int, n int) bool {
	for _, v := range r {
		if v[0] <= n && n <= v[1] {
			return true
		}
	}
	return false
}

func allEmpty(a [][]string) bool {
	for _, v := range a {
		if len(v) > 0 {
			return false
		}
	}
	return true
}

func contains(a []string, s string) (int, bool) {
	for i, v := range a {
		if v == s {
			return i, true
		}
	}
	return -1, false
}

func remove(s []string, r string) []string {
	for i, v := range s {
		if v == r {
			return append(s[:i], s[i+1:]...)
		}
	}
	return s
}
